### PR TITLE
[Model] Add PaliGemma

### LINKED ISSUE #4833: [New Model]: Google's Paligemma family of models
### The model to consider.

https://huggingface.co/google/paligemma-3b-pt-896

### The closest model vllm already supports.

I think the only visual language model supported right now is LLava but I could be wrong.

### What's your difficulty of supporting the model you want?

_No response_

### PR DESCRIPTION
**Pull request by ywang96**
Add support for [PaliGemma](https://huggingface.co/blog/paligemma) - Google's Cutting-Edge Open Vision Language Model.

FIX #4833

### CODE REVIEW

**Comment by ywang96** [comment]
Not ready for review yet but FYI @zhuohan123 @simon-mo since you asked me about this.

also cc @DarkLight1337 since I might need you to review this eventually

**Comment by ywang96** [comment]
Seeing some correctness issue with the implementation of `Gemma` itself - going to mark this PR back to draft for now.

File: tests/models/test_paligemma.py
Comment (DarkLight1337) [inline tests/models/test_paligemma.py:58]: Maybe indicate what this `108` means

**Comment by DarkLight1337** [comment]
Design-wise I think the code is fine. However, there appear to be major discrepancies between the expected and actual output beyond what can be observed for the base Gemma model. For example, the `caption es` example is not working at all (the model returns English output instead of Spanish).

**Comment by ywang96** [comment]
This PR is currently blocked by https://github.com/vllm-project/vllm/pull/5913 for end-to-end correctness.

**Comment by WoosukKwon** [comment]
@ywang96 Did you have a chance to test the model again after #5913?

**Comment by ywang96** [comment]
> @ywang96 Did you have a chance to test the model again after #5913?

@WoosukKwon I haven't, but I also need to update this PR because of the refactoring we just finished to make sure it works.

**Comment by ywang96** [comment]
This PR passed all tests locally and is ready for review. Please take a look @WoosukKwon!

File: vllm/model_executor/models/paligemma.py
Comment (DarkLight1337) [inline vllm/model_executor/models/paligemma.py:112]: I think the image token should already be omitted since `OpenAIServingChat.image_token_str` returns `None` for PaliGemma

File: vllm/model_executor/models/paligemma.py
Comment (DarkLight1337) [inline vllm/model_executor/models/paligemma.py:245]:
```suggestion
            vision_embeddings = vision_embeddings * self.config.hidden_size**-0.5
```

**Review by DarkLight1337 — COMMENTED** [review]
Overall LGTM but I'll check the warnings outputted in the VLM tests to see how consistent it is with the HF model.

File: vllm/model_executor/models/paligemma.py
Comment (ywang96) [reply vllm/model_executor/models/paligemma.py:112]: I don't think that's the case - it will still return `<image>`, but PaliGemma doesn't expect it to be in the prompt and processes it afterwards.

Comment (ywang96) [reply vllm/model_executor/models/paligemma.py:112]: nvm - I thought we didn't do it model by model there anymore, but that was getting the full vision language prompt, not the image token str

**Comment by DarkLight1337** [comment]
Seems that the vLLM output string is not following HF format:

https://buildkite.com/vllm/ci-aws/builds/4150#0190861d-cafe-41db-81ba-cb4c0d8fc2a6 

Can you fix that?

**Comment by DarkLight1337** [comment]
Seems that the vLLM output string is not following HF format:

https://buildkite.com/vllm/ci-aws/builds/4150#0190861d-cafe-41db-81ba-cb4c0d8fc2a6 

Can you fix that?

File: docs/source/models/supported_models.rst
Comment (DarkLight1337) [inline docs/source/models/supported_models.rst:134]:
```suggestion
  * - :code:`PaliGemmaForConditionalGeneration`
```

File: vllm/model_executor/models/paligemma.py
Comment (DarkLight1337) [reply vllm/model_executor/models/paligemma.py]: In that case this code should not be needed anymore. We can also log a warning if user tries to pass image tokens when no image is passed.

**Comment by ywang96** [comment]
> Seems that the vLLM output string is not following HF format:
> 
> https://buildkite.com/vllm/ci-aws/builds/4150#0190861d-cafe-41db-81ba-cb4c0d8fc2a6
> 
> Can you fix that?

Ah it's the eos_token. It's also weird that some of these tests actually don't give an output at all - I guess I will need to take a deeper look at it and see what went wrong.

File: docs/source/models/supported_models.rst
Comment (ywang96) [reply docs/source/models/supported_models.rst:134]: Will update this PR with the VLM doc PR when it's merged

File: vllm/model_executor/models/paligemma.py
Comment (ywang96) [reply vllm/model_executor/models/paligemma.py]: Yep I've updated with a warning to remove the user image token

File: docs/source/models/supported_models.rst
Comment (WoosukKwon) [inline docs/source/models/supported_models.rst:191]:
```suggestion
    - :code:`google/paligemma-3b-pt-224`, :code:`google/paligemma-3b-mix-224`, etc.
```

File: examples/paligemma_example.py
Comment (WoosukKwon) [inline examples/paligemma_example.py:1]: Do we need this example? I'm wondering because it seems pretty similar to the llava example.

File: tests/models/test_paligemma.py
Comment (WoosukKwon) [inline tests/models/test_paligemma.py:1]: I think this test is also a bit redundant with `test_llava.py`. Can we refactor `test_llava.py` to cover both models?

**Review by WoosukKwon — COMMENTED** [review]
@ywang96 Thanks for the PR! The PR looks good overall, but I think we can reduce the redundancy by reusing the code for Llava.

File: examples/paligemma_example.py
Comment (ywang96) [reply examples/paligemma_example.py:1]: The tests and examples were made for each model previously because the image-related engine args were different from each model. This is indeed no longer needed after the refactoring we did, and I will open another PR for consolidate them if you're okay with that.

File: tests/models/test_paligemma.py
Comment (ywang96) [reply tests/models/test_paligemma.py:1]: See my comment above.

File: examples/paligemma_example.py
Comment (WoosukKwon) [reply examples/paligemma_example.py:1]: Thanks for the explanation! Sounds good. Please open the PR!

**Review by WoosukKwon — APPROVED** [review]
@ywang96 LGTM. Please feel free to open the PR for refactoring.

Also, please merge the PR after @DarkLight1337 also approves as he might be more knowledgeable about the models than me.

**Comment by DarkLight1337** [comment]
Although the vLLM output for Spanish translation has some numerical difference and diverges from HF output after the first 2 tokens, I think the overall meaning is similar enough so it is fine. Let's merge this.

**Review by DarkLight1337 — APPROVED** [review]

**Pull request merged**
